nucliadb_node/nucliadb_node/shadow_shards.py

```diff
@@ -0,0 +1,253 @@
+        await self.save_metadata()
+        return shard_id
+
+    async def delete(self, shard_id: str) -> None:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+
+        if shard_id not in self.shards:
+            raise ShadowShardNotFound()
+
+        shard_path = self.shard_path(shard_id)
+        await aiofiles.os.remove(shard_path)
+        self.shards.remove(shard_id)
+
+        self.metadata.shards.pop(shard_id)
+        await self.save_metadata()
+
+    def exists(self, shard_id: str) -> bool:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        return shard_id in self.shards
+
+    def encode_operation(self, node_operation: NodeOperation) -> bytes:
+        opcode, payload = node_operation
+        if opcode == OperationCode.SET:
+            payload = payload.SerializeToString()  # type: ignore
+        elif opcode == OperationCode.DELETE:
+            payload = payload.encode()  # type: ignore
+        else:
+            ValueError(f"Unknown opcode: {opcode}")
+        encoded = base64.b64encode(payload).decode()  # type: ignore
+        return (opcode + encoded + "\n").encode()
+
+    def decode_operation(self, encoded: str) -> NodeOperation:
+        opcode = OperationCode(encoded[:4])
+        decoded_payload = base64.b64decode(encoded[4:])
+        if opcode == OperationCode.SET:
+            return (opcode, Resource.FromString(decoded_payload))
+        elif opcode == OperationCode.DELETE:
+            return (opcode, decoded_payload.decode())
+        raise ValueError(f"Unknown opcode: {opcode}")
+
+    async def set_resource(self, brain: Resource, shard_id: str) -> None:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        if not self.exists(shard_id):
+            raise ShadowShardNotFound()
+
+        encoded = self.encode_operation((OperationCode.SET, brain))
+        shard_path = self.shard_path(shard_id)
+        async with aiofiles.open(shard_path, mode="ab") as f:
+            await f.write(encoded)
+
+        await self.increment_ops(shard_id)
+
+    async def delete_resource(self, uuid: str, shard_id: str) -> None:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        if not self.exists(shard_id):
+            raise ShadowShardNotFound()
+
+        encoded = self.encode_operation((OperationCode.DELETE, uuid))
+        shard_path = self.shard_path(shard_id)
+        async with aiofiles.open(shard_path, mode="ab") as f:
+            await f.write(encoded)
+
+        await self.increment_ops(shard_id)
+
+    async def iter_operations(self, shard_id: str) -> AsyncIterator[NodeOperation]:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        if not self.exists(shard_id):
+            raise ShadowShardNotFound()
+
+        shard_path = self.shard_path(shard_id)
+        async with aiofiles.open(shard_path, "r") as f:
```

Comment:
Instead of reading files we could iterate with glob

Reply:
mh I am not sure what you mean. Here we are returning all the operations that were written in a particular shadow shard. So we need to read the file.